src/Interpreters/InterpreterCreateQuery.cpp

```diff
@@ -957,6 +957,19 @@ BlockIO InterpreterCreateQuery::createTable(ASTCreateQuery & create)
     if (!created)   /// Table already exists
         return {};
 
+    /// Check type compatible for materialized dest table and select columns
+    if (create.select && create.is_materialized_view && create.to_table_id)
+    {
+        StoragePtr table = DatabaseCatalog::instance().getTable({create.database, create.table, create.uuid}, getContext());
+        const auto & output_columns = table->getInMemoryMetadataPtr()->getSampleBlock();
```

Comment:
I think, here should be columns from `to_table`. You already mentioned some weired case, when not all columns from select query goes to materialized view, but all columns goes to destination table. Like this:
```sql
CREATE TABLE src (n UInt64, s FixedString(16)) ENGINE=Memory;
CREATE TABLE dst (n UInt8, s String) ENGINE = Memory;
CREATE MATERIALIZED VIEW mv TO dst (n String) AS SELECT * FROM src;
```

If you will use all columns from destination table, it will be enough to use just `InterpreterSelectQuery`.

Reply:
But if I use `InterpreterSelectQuery `:
```
StoragePtr table = DatabaseCatalog::instance().getTable({create.to_table_id.database_name, create.to_table_id.table_name, create.to_table_id.uuid}, getContext());
const auto & output_columns = table->getInMemoryMetadataPtr()->getSampleBlock();
Block input_columns = InterpreterSelectQuery(
            create.select->clone(), getContext(), SelectQueryOptions().analyze()).getSampleBlock();
```
It will throw error: `Logical error: 'Bad cast from type DB::ASTSelectWithUnionQuery to DB::ASTSelectQuery'`, because `create.select` is an `ASTSelectWithUnionQuery` not `ASTSelectQuery`. So should I convert `create.select` from `ASTSelectWithUnionQuery ` to `ASTSelectQuery ` first? 

What kind of extra works will `ASTSelectWithUnionQuery` do when compared to `ASTSelectWithUnionQuery `?

And I think we should check type for both columns of `to_table` and `materialized  view` with `select` columns.